kms/asymmetric/samples.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"golang.org/x/net/context"
+	"google.golang.org/api/cloudkms/v1"
+	"log"
+	"math/big"
+
+	"golang.org/x/oauth2/google"
+)
+
+// [START kms_get_asymmetric_public]
+// Retrieve a public key from a saved asymmetric key pair on KMS
+func getAsymmetricPublicKey(client *cloudkms.Service, ctx context.Context, keyPath string) (interface{}, error) {
+	response, err := client.Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions.
+		GetPublicKey(keyPath).Context(ctx).Do()
+	if err != nil {
+		return nil, err
+	}
+	if response == nil {
+		return nil, fmt.Errorf("no response from GetPublicKey")
+	}
+	keyBytes := []byte(response.Pem)
+	block, _ := pem.Decode(keyBytes)
+	publicKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	return publicKey, nil
+}
+
+// [END kms_get_asymmetric_public]
+
+// [START kms_decrypt_rsa]
+// Attempt to decrypt ciphertext with saved RSA key
+func decryptRSA(client *cloudkms.Service, ctx context.Context, ciphertext, keyPath string) (string, error) {
+	decryptRequest := &cloudkms.AsymmetricDecryptRequest{
+		Ciphertext: ciphertext,
+	}
+	response, err := client.Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions.
+		AsymmetricDecrypt(keyPath, decryptRequest).Context(ctx).Do()
+	if err != nil {
+		return "", err
+	}
+	messageArr, err := base64.StdEncoding.DecodeString(response.Plaintext)
+	if err != nil {
+		return "", err
+	}
+	message := fmt.Sprintf("%s", messageArr)
+	return message, nil
+}
+
+// [END kms_decrypt_rsa]
+
+// [START kms_encrypt_rsa]
+// Encrypt plaintext message using saved RSA public key
```

Comment:
`encryptRSA creates a ciphertext from a plain message using a RSA public key saved at the specified keyPath.`